exchanges/stream/websocket.go

```diff
@@ -376,63 +373,57 @@ func (w *Websocket) dataMonitor() {
 }
 
 // connectionMonitor ensures that the WS keeps connecting
-func (w *Websocket) connectionMonitor() error {
-	if w.checkAndSetMonitorRunning() {
-		return errAlreadyRunning
-	}
+func (w *Websocket) connectionMonitor() {
+	w.setConnectionMonitorRunning(true)
 	delay := w.connectionMonitorDelay
-
-	go func() {
-		timer := time.NewTimer(delay)
-		for {
+	timer := time.NewTimer(delay)
```

Comment:
Optionally you can apply this diff which removes the methods on methods approach:
```diff
diff --git a/exchanges/stream/websocket.go b/exchanges/stream/websocket.go
index 954c0c4ad..511ca12c0 100644
--- a/exchanges/stream/websocket.go
+++ b/exchanges/stream/websocket.go
@@ -34,7 +34,6 @@ var (

 // Private websocket errors
 var (
-       errAlreadyRunning                       = errors.New("connection monitor is already running")
        errExchangeConfigIsNil                  = errors.New("exchange config is nil")
        errExchangeConfigEmpty                  = errors.New("exchange config is empty")
        errWebsocketIsNil                       = errors.New("websocket is nil")
@@ -292,7 +291,7 @@ func (w *Websocket) Connect() error {
        }
        w.setState(connected)

-       if !w.IsConnectionMonitorRunning() {
+       if !w.connectionMonitorRunning.Swap(true) {
                go w.connectionMonitor()
        }

@@ -374,7 +373,6 @@ func (w *Websocket) dataMonitor() {

 // connectionMonitor ensures that the WS keeps connecting
 func (w *Websocket) connectionMonitor() {
-       w.setConnectionMonitorRunning(true)
        delay := w.connectionMonitorDelay
        timer := time.NewTimer(delay)
        for {
@@ -394,7 +392,7 @@ func (w *Websocket) connectionMonitor() {
                                log.Debugf(log.WebsocketMgr, "%v websocket: connection monitor exiting", w.exchangeName)
                        }
                        timer.Stop()
-                       w.setConnectionMonitorRunning(false)
+                       w.connectionMonitorRunning.Store(false)
                        return
                }
                select {
@@ -628,19 +626,6 @@ func (w *Websocket) IsTrafficMonitorRunning() bool {
        return w.trafficMonitorRunning.Load()
 }

-func (w *Websocket) checkAndSetMonitorRunning() (alreadyRunning bool) {
-       return !w.connectionMonitorRunning.CompareAndSwap(false, true)
-}
-
-func (w *Websocket) setConnectionMonitorRunning(b bool) {
-       w.connectionMonitorRunning.Store(b)
-}
-
-// IsConnectionMonitorRunning returns status of connection monitor
-func (w *Websocket) IsConnectionMonitorRunning() bool {
-       return w.connectionMonitorRunning.Load()
-}
-
 func (w *Websocket) setDataMonitorRunning(b bool) {
        w.dataMonitorRunning.Store(b)
 }
diff --git a/exchanges/stream/websocket_test.go b/exchanges/stream/websocket_test.go
index 6f150359c..701f2a8d0 100644
--- a/exchanges/stream/websocket_test.go
+++ b/exchanges/stream/websocket_test.go
@@ -571,13 +571,14 @@ func TestConnectionMonitor(t *testing.T) {
        ws.exchangeName = "hello"
        ws.Wg = &sync.WaitGroup{}
        ws.setEnabled(true)
+       ws.connectionMonitorRunning.Store(true)
        go ws.connectionMonitor()
        require.EventuallyWithT(t, func(c *assert.CollectT) {
-               assert.True(c, ws.IsConnectionMonitorRunning(), "IsConnectionMonitorRunning should return true")
+               assert.True(c, ws.connectionMonitorRunning.Load(), "IsConnectionMonitorRunning should return true")
        }, 5*time.Second, 10*time.Millisecond, "ConnectionMonitor must be running")
        ws.setEnabled(false)
        require.EventuallyWithT(t, func(c *assert.CollectT) {
-               assert.False(c, ws.IsConnectionMonitorRunning(), "IsConnectionMonitorRunning should return false")
+               assert.False(c, ws.connectionMonitorRunning.Load(), "IsConnectionMonitorRunning should return false")
        }, 5*time.Second, 10*time.Millisecond, "ConnectionMonitor must not be running")
 }
```

Reply:
Actually, I don't think we need an atomic at all.
* The only thing that starts it is Connect
* Connect is protected by a full struct mutex
* And bools are inherently atomic if you're not using compareAndSwap anyway, not that this matters given ☝️ 

I'm also curious about the state of any time we find we run Connect, it gets this far, but connectionMonitor is still running.
To me that implies that it's the *last* connectionMonitor, we're in a race, and it's about to exit.
Prove me wrong?

Reply:
Yeah good point, we can probably remove it then we might be able to add it to the waitgroup?

Reply:
Yep, that works best.
* Remove all "IsRunning" flags and checks
* Add to waitgroup
* Double check if this is already how the partner in crime `dataMonitor` works. They should follow the same pattern

Reply:
Agreed 🦅 Nice.

Reply:
I have removed the flag for the connection monitor and will cover dataMonitor in a separate PR. Is that ok?

Reply:
I think this is really a coherent chunk of work: Simplify Websocket monitor routines.
So should be one PR.